src/device/virtio/mod.rs

```diff
@@ -0,0 +1,159 @@
+// disabled. Let's figure out at some point if having MMIO as part of the name is necessary.
+const VIRTIO_MMIO_INT_VRING: u8 = 0x01;
+
+// The driver will write to the register at this offset in the MMIO region to notify the device
+// about available queue events.
+const VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET: u64 = 0x50;
+
+// TODO: Make configurable for each device maybe?
+const QUEUE_MAX_SIZE: u16 = 256;
+
+#[derive(Copy, Clone)]
+pub struct MmioConfig {
+    pub range: MmioRange,
+    // The interrupt assigned to the device.
+    pub gsi: u32,
+}
+
+// These arguments are common for all virtio devices. We're always passing a mmio_cfg object
+// for now, and we'll re-evaluate the layout of this struct when adding more transport options.
+pub struct CommonArgs<'a, M, B> {
+    // The objects used for guest memory accesses and other operations.
+    pub mem: M,
+    // Used by the devices to register ioevents and irqfds.
+    //pub vm_fd: Arc<VmFd>, FIXME remove
+    pub vmm: Arc<Hypervisor>,
+    // Mutable handle to the event manager the device is supposed to register with. There could be
+    // more if we decide to use more than just one thread for device model emulation.
+    pub event_mgr: &'a mut EventManager<Arc<Mutex<dyn MutEventSubscriber + Send>>>,
+    // This stands for something that implements `MmioManager`, and can be passed as a reference
+    // or smart pointer (such as a `Mutex` guard).
+    pub mmio_mgr: B,
+    // The virtio MMIO device parameters (MMIO range and interrupt to be used).
+    pub mmio_cfg: MmioConfig,
+    // We pass a mutable reference to the kernel cmdline `String` so the device can add any
+    // required arguments (i.e. for virtio over MMIO discovery). This means we need to create
+    // the devices before loading he kernel cmdline into memory, but that's not a significant
+    // limitation.
+    //pub kernel_cmdline: &'a mut Cmdline, FIXME remove
+}
+
+/// Simple trait to model the operation of signalling the driver about used events
+/// for the specified queue.
+// TODO: Does this need renaming to be relevant for packed queues as well?
+pub trait SignalUsedQueue {
+    // TODO: Should this return an error? This failing is not really recoverable at the interface
+    // level so the expectation is the implementation handles that transparently somehow.
+    fn signal_used_queue(&self, index: u16);
+}
+
+/// Uses a single irqfd as the basis of signalling any queue (useful for the MMIO transport,
+/// where a single interrupt is shared for everything).
+pub struct SingleFdSignalQueue {
+    pub irqfd: Arc<EventFd>,
+    pub interrupt_status: Arc<AtomicU8>,
+}
+
+impl SignalUsedQueue for SingleFdSignalQueue {
+    fn signal_used_queue(&self, _index: u16) {
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING, Ordering::SeqCst);
+        self.irqfd
+            .write(1)
+            .expect("Failed write to eventfd when signalling queue");
```

Comment:
I would log this error to not crash the system.

Reply:
done b84aa85